### PR TITLE
Add tss2-dlopen-esys.c

### PR DESCRIPTION
**Pull request by AndreasFuchsSIT**
Add a tss2-dlopen-esys that can be copied into project that do not want to link against libtss2-esys at compile time.
Same for tctildr, mu, rc and fapi.
Also added tests for esys, mu, rc, fapi and failure.

### CODE REVIEW

**Comment by AndreasFuchsSIT** [comment]
@poettering Would you like to review this as it might be an approach for simplifying https://github.com/systemd/systemd/blob/b69855e645a046d0860f2ae99464019e0f6adf81/src/shared/tpm2-util.c ?

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 1]
```diff
@@ -0,0 +1,2171 @@
...
+static void *esys_dlhandle = NULL;
+
+TSS2_RC
+Esys_Initialize(
+    ESYS_CONTEXT **esys_context,
+    TSS2_TCTI_CONTEXT *tcti,
+    TSS2_ABI_VERSION *abiVersion)
+{
+    if (!esys_dlhandle) {
+        esys_dlhandle = dlopen("tss2_esys", RTLD_NOW & RTLD_LOCAL);
```
Comment (poettering) [inline tss2-dlopen/tss2-dlopen-esys.c:34]: i figure you mean | rather than & here...

**Review by poettering — CHANGES_REQUESTED** [review]

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 2]
```diff
@@ -0,0 +1,2171 @@
...
+    if (!sym) {
+        fprintf(stderr, "WARNING: Function Esys_Finalize not found.");
+        return;
+    }
+    sym(ctx);
+}
+
+void
+Esys_Free(void *__ptr)
+{
```
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply tss2-dlopen/tss2-dlopen-esys.c:63]: Thanks for the review. You mean putting `if (!__ptr) return;` to the beginning of the function ?

[the remark this answers is not in the document]
Comment (poettering) [reply tss2-dlopen/tss2-dlopen-esys.c:63]: yeah. precisely.

And Esys_Finalize() should probably do the same:

```c
if (!*ctx)  {
        *ctx = NULL;
        return;  
}
```

**Comment by AndreasFuchsSIT** [comment]
Force-Push with Lennart's comments (hopefully) addressed.
If Peter and Bill are ok too, I'll go ahead and also make wrappers for tctildr, mu, rc and fapi.

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 3]
```diff
@@ -0,0 +1,2171 @@
...
+TSS2_RC
+Esys_Initialize(
+    ESYS_CONTEXT **esys_context,
+    TSS2_TCTI_CONTEXT *tcti,
+    TSS2_ABI_VERSION *abiVersion)
+{
+    if (!esys_dlhandle) {
+        esys_dlhandle = dlopen("tss2_esys", RTLD_NOW & RTLD_LOCAL);
+        if (!esys_dlhandle) {
+            fprintf(stderr, "WARNING: Library libtss2_esys not found.");
```
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply tss2-dlopen/tss2-dlopen-esys.c]: Wrapped into a ifdef'd macro

File: tss2-dlopen/tss2-dlopen-esys.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply tss2-dlopen/tss2-dlopen-esys.c:75]: Without the `*ctx=NULL` NOP, but yes, done... ;-)
Thanks !

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 4]
```diff
@@ -0,0 +1,2171 @@
...
+    }
+    sym(__ptr);
+}
+
+#define MAKE_ESYS_0(fun) \
+TSS2_RC fun (ESYS_CONTEXT *ctx) { \
+    TSS2_RC (*sym) (ESYS_CONTEXT *ctx) = \
+        dlsym(esys_dlhandle, str(fun)); \
+    if (!sym) { \
+        fprintf(stderr, "WARNING: Function " str(fun) " not found."); \
```
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply tss2-dlopen/tss2-dlopen-esys.c]: Does it really matter ?
I just took the same string was for symbol lookup.

Could be `__func__` there as well, I agree but maybe someone wants to rename the function names to eg `dl_ ## fun` and then it would not be correct here anymore.

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 5]
```diff
@@ -0,0 +1,2201 @@
...
+ * For new versions of this file, please check:
+ * http://github.com/tpm2-software/tpm2-tss/tss2-dlopen
+*/
+
+#include <dlfcn.h>
+#include <stdio.h>
+#include <tss2/tss2_esys.h>
+
+#define str(s) xstr(s)
+#define xstr(s) #s
```
Comment (PeterHuewe) [inline tss2-dlopen/tss2-dlopen-esys.c:23]: don't get it - why is xstr needed?

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 6]
```diff
@@ -0,0 +1,2201 @@
...
+    if (!sym) \
+        dlsym(esys_dlhandle, str(fun)); \
+    if (!sym) { \
+        WARN("Function " str(fun) " not found."); \
+        return TSS2_ESYS_RC_NOT_IMPLEMENTED; \
+    } \
+    return sym(ctx, parm1, parm2, parm3, parm4); \
+}
+
+#define MAKE_ESYS_5(fun, type1,parm1, type2,parm2, type3,parm3, type4,parm4, \
```
Comment (PeterHuewe) [inline tss2-dlopen/tss2-dlopen-esys.c:148]: for the other macros e.g. MAKE_ESYS_4 you don't "group" type,param but skip the spaces alltogether.
please make this consisten.

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 7]
```diff
@@ -0,0 +1,2201 @@
...
+MAKE_ESYS_1(Esys_GetTcti,
+    TSS2_TCTI_CONTEXT **,tcti);
+MAKE_ESYS_2(Esys_GetPollHandles,
+    TSS2_TCTI_POLL_HANDLE **,handles,
+    size_t *,count);
+MAKE_ESYS_1(Esys_SetTimeout,
+    int32_t, timeout);
+MAKE_ESYS_3(Esys_TR_Serialize,
+    ESYS_TR, object,
+    uint8_t **,buffer,
```
Comment (PeterHuewe) [inline tss2-dlopen/tss2-dlopen-esys.c:307]: same here, sometimes you add spaces, sometimes you don't - please make consistent.

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 8]
```diff
@@ -0,0 +1,2201 @@
...
+    if (!sym) { \
+        WARN("Function " str(fun) " not found."); \
+        return TSS2_ESYS_RC_NOT_IMPLEMENTED; \
+    } \
+    return sym(ctx, parm1, parm2, parm3, parm4, parm5, parm6, parm7, parm8, \
+                    parm9, parm10, parm11, parm12, parm13); \
+}
+
+
+
```
Comment (PeterHuewe) [inline tss2-dlopen/tss2-dlopen-esys.c:297]: what is the order of these functions below?

**Comment by PeterHuewe** [comment]
Few nitpicks, but looks okay to me.

File: tss2-dlopen/tss2-dlopen-esys.c (hunk 8 above)
Comment (AndreasFuchsSIT) [reply tss2-dlopen/tss2-dlopen-esys.c:297]: The order is the same as in the header file which is the same as in the TPM spec part 3.

File: tss2-dlopen/tss2-dlopen-esys.c (hunk 5 above)
Comment (AndreasFuchsSIT) [reply tss2-dlopen/tss2-dlopen-esys.c:23]: It's a best practice that has to do with order of macro expansion. You want to expand the contents of s before you stringify it. Don't ask me exactly why, but I learned it somewhere... ;-)

Comment (PeterHuewe) [reply tss2-dlopen/tss2-dlopen-esys.c:23]: ok

File: tss2-dlopen/tss2-dlopen-esys.c (hunk 6 above)
Comment (PeterHuewe) [reply tss2-dlopen/tss2-dlopen-esys.c:148]: resolved but not changed?

File: tss2-dlopen/tss2-dlopen-esys.c (hunk 8 above)
Comment (PeterHuewe) [reply tss2-dlopen/tss2-dlopen-esys.c]: ok

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 9]
```diff
@@ -0,0 +1,2199 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*******************************************************************************
+ * Copyright 2021, Fraunhofer SIT
+ * All rights reserved. *******************************************************************************/
```
Comment (poettering) [inline tss2-dlopen/tss2-dlopen-esys.c:4]: hmm, wasn#t there supposed to be one more line break in this license header?

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 10]
```diff
@@ -0,0 +1,2199 @@
...
+
+#include <dlfcn.h>
+#include <stdio.h>
+#include <tss2/tss2_esys.h>
+
+#define str(s) xstr(s)
+#define xstr(s) #s
+
+#ifdef ENABLE_WARN
+#define WARN(str, ...) fprintf(stderr, "WARNING: " str "\n", ## __VA_ARGS__);
```
Comment (poettering) [inline tss2-dlopen/tss2-dlopen-esys.c:26]: if you define it this way, then you will insert two ";" in the code. Which is a bit weird. just drop the ";"?

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 11]
```diff
@@ -0,0 +1,2199 @@
...
+#else /* ENABLE_WARN */
+#define WARN(...) ;
```
Comment (poettering) [inline tss2-dlopen/tss2-dlopen-esys.c:28]: change do: `do {} while(0)` or so?

That way people can put a semicolon at the end and will not get the right number of expressions and semicolons, so that you can then do this safely:

```c
if (x)
        WARN(…);
```

And you'll get exactly one ";" generated and no uncertainty what is done as part of the if branch and what not

File: tss2-dlopen/tss2-dlopen-mu.c [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+#ifdef ENABLE_WARN
+#define WARN(str, ...) fprintf(stderr, "WARNING: " str "\n", ## __VA_ARGS__);
+#else /* ENABLE_WARN */
+#define WARN(...) ;
+#endif /* ENABLE_WARN */
+
+static void *mu_dlhandle = NULL;
+
+static TSS2_RC
+init_dlhandle()
```
Comment (poettering) [inline tss2-dlopen/tss2-dlopen-mu.c:34]: this isn't valid C. `init_dlhandle()` → `init_dlhandle(void)`

File: tss2-dlopen/tss2-dlopen-tctildr.c [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+#ifdef ENABLE_WARN
+#define WARN(str, ...) fprintf(stderr, "WARNING: " str "\n", ## __VA_ARGS__);
+#else /* ENABLE_WARN */
+#define WARN(...) ;
+#endif /* ENABLE_WARN */
+
+static void *tctildr_dlhandle = NULL;
+
+static TSS2_RC
+init_dlhandle()
```
Comment (poettering) [inline tss2-dlopen/tss2-dlopen-tctildr.c:34]: ditto

File: Makefile.am [hunk 14]
```diff
@@ -219,6 +219,14 @@ EXTRA_DIST += \
     src/tss2-tcti/tss2-tcti-tbs.vcxproj.filters \
     tpm2-tss.sln
 
+# tss2-dlopen wrappers
+EXTRA_DIST += \
+    tss2-dlopen/tss2-dlopen-rc.c \
+    tss2-dlopen/tss2-dlopen-tctildr.c \
+    tss2-dlopen/tss2-dlopen-mu.c \
+    tss2-dlopen/tss2-dlopen-esys.c \
+    tss2-dlopen/tss2-dlopen-fapi.c
```
Comment (williamcroberts) [inline Makefile.am:228]: i'm fine with this, just need tests and negative test (cmocka).

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 15]
```diff
@@ -0,0 +1,2199 @@
...
+        }
+    }
+
+    static TSS2_RC (*sym) (ESYS_CONTEXT **esys_context, TSS2_TCTI_CONTEXT *tcti, TSS2_ABI_VERSION *abiVersion) = NULL;
+    if (!sym)
+        sym = dlsym(esys_dlhandle, "Esys_Initialize");
+    if (!sym) {
+        WARN("Function Esys_Initialize not found.");
+        return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+    }
```
Comment (williamcroberts) [inline tss2-dlopen/tss2-dlopen-esys.c:54]: personally, I would do all the heavy lifting in `Esys_Initialize()` and do something like:
```C
malloc(esys_hdr + sizeof(function table);
```
So your structure for esys ctx pointer is:
```
Normal esys ctx
function table
```

Then in your subsequent calls no dlsym needed, you just:
```
if (esys_ctx) {
   esys_ctx->fn1()
}
```
However that would mean esys ctx would be incompatible between dlopen and standard versions, but I don't know how they would end up with both symbol names in their program without doing something weird to bring it on themselves. You could pass a dl-open ctx to the lib version but not vice versa.

Comment (williamcroberts) [reply tss2-dlopen/tss2-dlopen-esys.c:54]: I guess you could just make the whole fn table static global and initialize it in Esys_Initialize, but you would want to make it thread safe, which it currently isn't with your use of static, but the dlopen/dlsym/dlclose are MT safe. However, the scenario of N>1 threads updating the sym variable which are not atomic.

Comment (AndreasFuchsSIT) [reply tss2-dlopen/tss2-dlopen-esys.c:63]: Is it really worth moving all the dlsym() calls into the intiialize ?
It will end up being slower because we make a lot more dlsym() calls than the application really needs. So I see no benefit here.

Wrt being thread safe: The pointer writes are atomic and calling dlsym twice will also not be a problem. If we end up with two handles to libesys / etc it's also ok, because the libtss2_* do not have global state so we can even mix and mingle between those.
Thus, I'd keep the current lightweight version.
If people want to do it differently when they copy these files, they can of course rewrite them to their liking.

File: tss2-dlopen/tss2-dlopen-esys.c [hunk 16]
```diff
@@ -0,0 +1,2208 @@
...
+#include <stdio.h>
+#include <tss2/tss2_esys.h>
+
+#define str(s) xstr(s)
+#define xstr(s) #s
+
+#ifdef ENABLE_WARN
+#define WARN(str, ...) do { fprintf(stderr, "WARNING: " str "\n", ## __VA_ARGS__); } while (0)
+#else /* ENABLE_WARN */
+#define WARN(...) do { ; } while (0)
```
Comment (poettering) [inline tss2-dlopen/tss2-dlopen-esys.c:29]: you can drop the ";"...

**Comment by AndreasFuchsSIT** [comment]
@williamcroberts I cannot reproduce the asan failure locally. Do you have any clue what that even means ?
```
FAIL: test/unit/dlopen-fail
===========================

==100267==AddressSanitizer CHECK failed: ../../../../src/libsanitizer/asan/asan_interceptors.cc:734 "((__interception::real_memcpy)) != (0)" (0x0, 0x0)
    <empty stack>

FAIL test/unit/dlopen-fail (exit status: 1)
```

**Comment by williamcroberts** [comment]
> "((__interception::real_memcpy)) != (0)" (0x0, 0x0)

no idea, can't you fetch the Docker image and run in there, does it work with newer versions of clang?

I have these aliases in my bashrc:
```bash
_docker_run() {
  local b=$(basename $(pwd)) 
  #docker run --cap-add=SYS_PTRACE --ulimit core=0 -it --env-file .ci/docker.env -v "`pwd`:/workspace/$b" "$1" /bin/bash
  docker run --cap-add=SYS_PTRACE --memory=14GB --cpus=2 -it --env-file .ci/docker.env -v "`pwd`:/workspace/$b" "$1" /bin/bash
}

docker_run() {
  local r="ghcr.io/tpm2-software/${1:-ubuntu-16.04}"
  _docker_run "$r"
}
```

then I cd to the project:
```bash
cd ~/workspace/tpm2-tss
```

Then I run one of the containers:
```bash
docker_run ubuntu-18.04
```

Then you can do something like setting the env. you could optionally do these things before the `docker_run`.

```bash
export CC=clang
```

Then run the ci script or individual commands.

**Comment by AndreasFuchsSIT** [comment]
For the record: The reason was that I did implement dlopen() and dlsym() instead of -ldl.
Apparently libasan has problems with that. So you always have to --wrap=dlopen/sym even if you don't link against -ldl in your test.
Just keeping this info here for the record.

File: tss2-dlopen/tss2-dlopen-esys.c (hunk 15 above)
Comment (williamcroberts) [reply tss2-dlopen/tss2-dlopen-esys.c:63]: Its MT-Safe so we're fine here.

**Review by williamcroberts — APPROVED** [review]
